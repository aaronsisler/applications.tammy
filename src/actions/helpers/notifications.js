export const setNotifications = (notifications) => ({
    type: 'SET_NOTIFICATIONS',
    notifications,
});
