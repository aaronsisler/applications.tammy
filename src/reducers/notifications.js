const notificationsReducerDefaultState = [];

const notificationsReducer = (state = notificationsReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_NOTIFICATIONS':
            return action.notifications;
        default:
            return state;
    }
};

export default notificationsReducer;

