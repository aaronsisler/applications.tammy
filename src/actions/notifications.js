import database from 'Firebase/firebase';
import { setNotifications } from 'Actions/helpers/notifications';

export const startStreamNotifications = () => (dispatch, getState) => {
    const { uid: userId } = getState().auth;

    return database.ref(`notifications/${userId}`).on('value', (snapshot) => {
        const notifications = [];
        snapshot.forEach((childSnapshot) => {
            notifications.push({
                notificationId: childSnapshot.key,
                ...childSnapshot.val(),
            });
        });

        return dispatch(setNotifications(notifications));
    });
}
