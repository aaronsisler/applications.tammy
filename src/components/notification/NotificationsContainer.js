import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { startStreamNotifications } from 'Actions/notifications';
import NotificationsIcon from 'Notification/NotificationsIcon';
import NotificationsList from 'Notification/NotificationsList';

export class NotificationsContainer extends React.Component {
    constructor(props) {
        super(props);
        this.props.startStreamNotifications();
        this.state = {
            isListToggledOpen: false
        }
    }

    handleCountUnreadNotifications = () => {
        const { notifications } = this.props;
        let unreadNotificationsCount = 0;
        notifications.map((notification) => notification.hasBeenViewed ? undefined : unreadNotificationsCount++);
        return unreadNotificationsCount;
    }

    handleToggleNotificationsList = () => this.setState((prevState) => ({
        isListToggledOpen: !prevState.isListToggledOpen,
    }))

    render() {
        const unreadNotificationsCount = this.handleCountUnreadNotifications();
        return (
            <div id="notifications_container">
                <div
                    className="notifications_container__wrapper"
                    onClick={this.handleToggleNotificationsList}
                >
                    <NotificationsIcon unreadNotificationsCount={unreadNotificationsCount} />
                </div>
                {this.state.isListToggledOpen && <NotificationsList notifications={this.props.notifications} />}
            </div>
        )
    }
}

/* istanbul ignore next */
const mapStateToProps = (state) => ({
    notifications: state.notifications,
})

/* istanbul ignore next */
const mapDispatchToProps = (dispatch) => ({
    startStreamNotifications: () => dispatch(startStreamNotifications()),
})

export default connect(mapStateToProps, mapDispatchToProps)(NotificationsContainer);

NotificationsContainer.propTypes = {
    notifications: PropTypes.array,
    startStreamNotifications: PropTypes.func.isRequired,
};
