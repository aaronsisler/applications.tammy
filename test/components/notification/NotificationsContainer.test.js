import React from 'react';
import { shallow } from 'enzyme';
import { NotificationsContainer } from 'Notification/NotificationsContainer';
import notifications from '../../fixtures/notifications';

describe('NotificationsContainer', () => {
    const startStreamNotifications = jest.fn();
    let wrapper;

    const buildWrapper = (notificationsInput = []) => {
        wrapper = shallow(
            <NotificationsContainer
                notifications={notificationsInput}
                startStreamNotifications={startStreamNotifications}
            />
        );
    };

    beforeEach(() => {
        buildWrapper(notifications);
    });

    describe('when is being constructed', () => {
        it('should set isListToggledOpenrender to false', () => {
            expect(wrapper.state('isListToggledOpen')).toBe(false);
        });

        it('should call startStreamNotifications', () => {
            expect(startStreamNotifications).toHaveBeenCalled();
        });
    });

    describe('when list is currently toggled closed', () => {
        it('should render NotificationsContainer correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });

        describe('when list is toggled to open', () => {
            it('should set the isListToggledOpen to false', () => {
                wrapper.find('.notifications_container__wrapper').simulate('click');

                expect(wrapper.state('isListToggledOpen')).toBe(true);
            });
        });
    });

    describe('when list is currently toggled open', () => {
        beforeEach(() => {
            wrapper.setState({ isListToggledOpen: true });
        });

        it('should render NotificationsContainer correctly', () => {
            wrapper.find('.notifications_container__wrapper').simulate('click');

            expect(wrapper).toMatchSnapshot();
        });

        describe('when list is toggled to close', () => {
            it('should set the isListToggledOpen to false', () => {
                wrapper.find('.notifications_container__wrapper').simulate('click');

                expect(wrapper.state('isListToggledOpen')).toBe(false);
            });
        });
    });
});
