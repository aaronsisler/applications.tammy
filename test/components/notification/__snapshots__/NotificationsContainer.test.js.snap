// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsContainer when list is currently toggled closed should render NotificationsContainer correctly 1`] = `
<div>
  <div
    id="notifications_container"
  >
    <div
      className="notifications_container__wrapper"
      onClick={[Function]}
    >
      <NotificationsIcon
        unreadNotificationsCount={1}
      />
    </div>
  </div>
</div>
`;

exports[`NotificationsContainer when list is currently toggled open should render NotificationsContainer correctly 1`] = `
<div>
  <div
    id="notifications_container"
  >
    <div
      className="notifications_container__wrapper"
      onClick={[Function]}
    >
      <NotificationsIcon
        unreadNotificationsCount={1}
      />
    </div>
  </div>
</div>
`;
